### DISCUSSION TITLE
Performance of OperatingSystem.getProcesses() over a large period of time

### DISCUSSION THREAD

**Discussion by unknown-spirit** [oshi/oshi#2073] [Q&A] [answered]
Hi guys,
I am using oshi for my [BusyTasks](https://www.pling.com/p/1201835/) app on Windows and just did some profiling to find out how to improve cpu performance.
I am calling getProcesses() every second once to gather all necessary information. Unfortunately, this method takes about 96,5% of all the cpu of my program, so it is wise to concentrate any performance tweak on that part. More precisely, I found that each call takes 75-100 ms at startup, which is quite acceptable. However, this value increases over time. After about 5 hours, it is already up to 200-250 ms. Likewise, the CPU that it takes on my system starts at around 2% and finishes at 6-7%. When I concurrently launch my program again and compare the results to the process running 5 hours, I see that they are surprisingly equal. It is only when I close both instances and restart the app that the values are back at initially low level. As a concrete interpretation I blame the JVM for the performance loss and wonder if there's actually anything that can be done to get a stable performance.
I appreciate any ideas and thoughts about it.

**Comment by dbwiddis** [comment, marked as answer]
Just a hunch, this may be related to finalizer usage, see #2009.  

You can try to inspect the heap when it gets slow to see if you've got GC issues.   

If this is the case, I'll be making some changes when JNA 5.12.0 is released (hopefully this week) and releasing OSHI 6.2.0 which should significantly improve performance.

If you want to pre-test a fraction of the possible improvements, check out this branch that I put in a portion of the expected 5.12.0 changes.  https://github.com/dbwiddis/oshi/tree/cleanermemory

**Reply by unknown-spirit** [reply]
Thanks for your support.
First of all, I want to correct CPU values. If I remember correctly this time, getProcesses() takes about 70% of the CPU of my program, not 96.5% as stated in the opener. I just got confused with the vast size of profiler values. Anyway, it's still much for a single call and thus worth looking into it. I'll have another profiling session later in the evening and may post a screenshot to give you some insight.

> Just a hunch, this may be related to finalizer usage, see https://github.com/oshi/oshi/issues/2009.

Interesting. Luckily, memory consumption of BusyTasks was steady over many hours though. Unfortunately I can't run the program for so many days like the opener of the referenced thread.

> You can try to inspect the heap when it gets slow to see if you've got GC issues.

I'll try to do so. For the record, it just does not suddenly get slow, but slowly increases over time.

What I still don't really understand is the fact that another fresh instance of BusyTasks performs as bad as the originally supervised and long-running one, as long as both are running. Does that mean that the JVM merges calls from different processes for performance reasons?

**Reply by dbwiddis** [reply]
Again, I'm mostly guessing here!

My suspicion is that the JVM is spending more time with "cleanup" tasks over time, and a mostly-full heap is being shared by the two processes.   I may be wrong here but I don't have any other ideas.

Investigating the heap using JVisualVM or other tools (see https://www.baeldung.com/java-heap-dump-capture) can help confirm this guess.

**Reply by unknown-spirit** [reply]
Hi,
I just did a more complete research with heap dumps at startup and after four hours, but all in all it just backs up my initial results.
As far as I see there's nothing suspicious going on. The heap increased from 10mb to 12mb, but the proportion of elements is similar.
As I am anything but an expert in computer science, it's quite hard for me to interpret all those collected values.

![Alt text](https://i.ibb.co/PFs1PsV/Visual-VM1.jpg "Optional title")

On the following screenshot you see the cpu time of the thread "Timer-1" that is periodically calling getProcesses(), it averages at around 220ms after four hours, but started at 75ms. As I write this, I see it even went up to 300ms. Unfortunately I cannot run a longer test interval, but I expect further and further increase.

![Alt text](https://i.ibb.co/tYtry1g/VisualVM.jpg "Optional title")

The new instance of BusyTasks coincides with these values until I shut everything down. Nothing new, I am afraid.

**Reply by dbwiddis** [reply]
Actually there's quite a bit there that indicates my hunch may be right.

![image](https://user-images.githubusercontent.com/9291703/174686678-5956bf9f-33ec-4e2e-8e33-91e57f6d72fd.png)

Some of the biggest offending classes on the heap are associated with `Memory` and the finalizer.  In the current state, when Memory becomes unreachable it's GC'd but the first run has to release memory with the finalizer so it's moved to survivor space. Once that's full it ends up in OldGen space where it's not addressed until stop-the-world GC.  And that may be happening more often the longer your program runs.

You can try experimenting with [GC-related switches](https://www.graalvm.org/22.1/reference-manual/native-image/MemoryManagement/) to see if they have any impact, perhaps trying `--gc=G1` first, and then possibly playing with heap size and various percentages.

Ultimately if you can diagnose this as the issue then the best thing to do is wait a week or so for the new JNA release and some optimization in OSHI to take advantage of it.

**Reply by dbwiddis** [reply]
So digging into this, I noticed the memory sizing in the HkeyPerformanceDataUtil cited on your "Dominators" list in the lower right was very inefficient. It did a trial-and-error sizing and resizing in units of pages every time.

On my VM with only 73 processes, it required 98304 bytes (24 4096-byte pages).  So that means every second, it was creating 24 times the memory allocations it needed.  On systems with more processes that can be even larger!

I made a change in #2075 to keep the max value and use it the next time, which should make a big improvement.  It just got merged and should be in 6.2.0-SNAPSHOT shortly.... can you test that out to see if it helps?

I expect even more improvement when the next JNA release comes out in about a week.

**Reply by unknown-spirit** [reply]
> I made a change in https://github.com/oshi/oshi/pull/2075 to keep the max value and use it the next time, which should make a big improvement. It just got merged and should be in 6.2.0-SNAPSHOT shortly.... can you test that out to see if it helps?

I just ran it and really realised a big difference. The CPU time of TIMER-1 started at about 30ms and just increased to about 60ms in two hours. I'll need more profiling time to say how it behaves on long term, but it's already an unbelievable change. Thanks already!

> You can try experimenting with [GC-related switches](https://www.graalvm.org/22.1/reference-manual/native-image/MemoryManagement/) to see if they have any impact, perhaps trying --gc=G1 first, and then possibly playing with heap size and various percentages.

Good idea, I may do so in future when I see that the issue is not yet solved.

**Reply by dbwiddis** [reply]
@unknown-spirit JNA 5.12.0 was released today and I just updated OSHI to depend on it and also leverage the Closeable memory.  I expect if you were to test again with the current snapshot you'd see even more improvement over the last time.   

Expect a 6.2.0 release soon.

**Reply by unknown-spirit** [reply]
Hi,
Thanks for the info.
The overall cpu time of my app and in particular of the process-fetching timer is quite low since the last change. Sadly, I encountered another problem. One of the basic system processes (ntoskrnl) is obviously interfering with my program and took all the CPU that it gave away. It gradually increases over time and was up to 10% after six hours! Once again, after closing BusyTasks, it faded away to 0.5%. I don't really have a clue what is going on here but there could be a connection to oshi. I will do more profiling to back up my observations.
I'm sure I'm going to try the new oshi-version too in the next days and let you know about further key findings!

**Reply by dbwiddis** [reply]
I suspect you're still seeing an impact from Garbage Collection, but not necessarily from OSHI.  Where are all the `byte[]` allocations coming from?

**Reply by unknown-spirit** [reply]
@dbwiddis
I decided to try the new oshi with new jna before digging into it any further, but it fails on runtime:

`Exception in thread "Timer-1" com.sun.jna.platform.win32.PdhUtil$PdhException: Pdh call failed with error code 0xC0000BB8`
        `at com.sun.jna.platform.win32.PdhUtil.PdhEnumObjectItems(PdhUtil.java:173)`
        `at oshi.util.platform.windows.PerfCounterWildcardQuery.queryInstancesAndValuesFromPDH(PerfCounterWildcardQuery.java:129)`

As far as I can tell this is a jna-problem, so I guess I should open an issue there?

**Reply by dbwiddis** [reply]
Rut Roh.  I updated that JNA method in the latest release, see https://github.com/java-native-access/jna/pull/1442.  I hope I didn't mess something up.  Can you post the full stack trace please?

**Reply by unknown-spirit** [reply]
Sure:

`Exception in thread "Timer-1" com.sun.jna.platform.win32.PdhUtil$PdhException: Pdh call failed with error code 0xC0000BB8`
`at com.sun.jna.platform.win32.PdhUtil.PdhEnumObjectItems(PdhUtil.java:173)`
`at oshi.util.platform.windows.PerfCounterWildcardQuery.queryInstancesAndValuesFromPDH(PerfCounterWildcardQuery.java:129)`
`at oshi.util.platform.windows.PerfCounterWildcardQuery.queryInstancesAndValues(PerfCounterWildcardQuery.java:86)`
`at oshi.driver.windows.perfmon.ProcessorInformation.queryProcessorCounters(ProcessorInformation.java:167)`
`at oshi.hardware.platform.windows.WindowsCentralProcessor.queryProcessorCpuLoadTicks(WindowsCentralProcessor.java:352)`
`at oshi.util.Memoizer$1.get(Memoizer.java:87)`
`at oshi.hardware.common.AbstractCentralProcessor.getProcessorCpuLoadTicks(AbstractCentralProcessor.java:204)`
`at oshi.hardware.platform.windows.WindowsCentralProcessor.querySystemCpuLoadTicks(WindowsCentralProcessor.java:221)`
`at oshi.util.Memoizer$1.get(Memoizer.java:87)`
`at oshi.hardware.common.AbstractCentralProcessor.getSystemCpuLoadTicks(AbstractCentralProcessor.java:192)`
`at BusyTasks.BusyTasks.zeigeCPUProzess(BusyTasks.java:924)`
`at BusyTasks.BusyTasks$9.run(BusyTasks.java:774)`
`at java.base/java.util.TimerThread.mainLoop(Timer.java:556)`
`at java.base/java.util.TimerThread.run(Timer.java:506)`

**Reply by dbwiddis** [reply]
The good news (?) is that the error line number (173) is before my recent changes so it's not associated with that PR.  I'm curious why it would just start now and not have been a problem before the update.

This is a common error in non-English installations if you have corrupted performance counters in the registry.  Often `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Perflib\009\Counter` is missing.  Can you investigate whether that key exists in your registry and if not, rebuild your performance counters?

**Reply by dbwiddis** [reply]
https://docs.microsoft.com/en-us/troubleshoot/windows-server/performance/rebuild-performance-counter-library-values

**Reply by unknown-spirit** [reply]
I just checked and saw that it exists.
And I just checked, jna 5.11 works fine.

**Reply by dbwiddis** [reply]
JNA 5.11 shouldn't work with the newest version of OSHI.  The previous version caught the error and fell back to WMI...

**Reply by dbwiddis** [reply]
So in previous versions of OSHI that exception was caught and falls back to WMI (which may be the cause of your high System load, by the way).  When backing out the workaround I had for the JNA issue, I did not put back the exception catch.   I've just fixed that in #2088, so at least SNAPSHOT will remove that exception and add a more useful log warning message.

There is still an issue finding the "Processor Information" value in those counters. Have you tried rebuilding them?

**Reply by dbwiddis** [reply]
Also when it "works" are there any log messages associated with performance counters?

**Reply by unknown-spirit** [reply]
>There is still an issue finding the "Processor Information" value in those counters. Have you tried rebuilding them?

I think you are right, there is a problem to find the "Processor Information" in the registry. It's just that I am hestitating to rebuild them, cause it seems like some vital stuff for windows.

**Reply by dbwiddis** [reply]
It is vital stuff which is why if it's corrupt you should fix it.    This is a relatively common occurrence and the instructions are simple. :)

**Reply by dbwiddis** [reply]
Just try these two commands in a command prompt (may need to be admin level):
```
cd \windows\system32
lodctr /R
```

**Reply by unknown-spirit** [reply]
> Just try these two commands in a command prompt:

They finished successfully. But this appears to be only the last step in the list.

**Reply by dbwiddis** [reply]
They should work if you don't have any custom counters.  See if it fixes the problem :)

**Reply by dbwiddis** [reply]
(Reboot first)

**Reply by unknown-spirit** [reply]
> They should work if you don't have any custom counters. See if it fixes the problem :)

Unfortunately it didn't. Rebuilt successfully and rebooted twice, but no difference.

**Reply by dbwiddis** [reply]
Using the perfmon.exe tool, can you see the "Processor Information" counters?

![image](https://user-images.githubusercontent.com/9291703/176034601-4666f2a8-8673-41dd-979e-84352f8d6e62.png)

**Reply by unknown-spirit** [reply]
Yes I can! They are not in English, though.

**Reply by dbwiddis** [reply]
OK, at the registry key mentioned earlier, can you see the "Processor Information" string at all?   `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Perflib\009\Counter`

See [this question](https://superuser.com/questions/453886/perfmon-counters-missing): note that both `/r` and `/R` are referenced.... I think it's supposed to be uppercase.  It may be you need to do the full extraction of the Perfh009.dat file to fix this.

**Reply by unknown-spirit** [reply]
>OK, at the registry key mentioned earlier, can you see the "Processor Information" string at all?

Yes, it is there.

>See [this question](https://superuser.com/questions/453886/perfmon-counters-missing): note that both /r and /R are referenced.... I think it's supposed to be uppercase. It may be you need to do the full extraction of the Perfh009.dat file to fix this.

You mean as mentioned [here](https://docs.microsoft.com/en-us/troubleshoot/windows-server/performance/rebuild-performance-counter-library-values)?

**Reply by dbwiddis** [reply]
Yeah, although that requires the install CD so not even sure where that is!  :( 

Let me investigate, it may be that those values are supposed to be localized?  This has always been hard for me to understand the API and troubleshoot this without a non-English system.

**Reply by unknown-spirit** [reply]
> Yeah, although that requires the install CD so not even sure where that is! :(

It's kinda sad, because even if I get it to work, I can't expect my users to go through that procedure just to run BusyTasks.

**Reply by dbwiddis** [reply]
Well, it may be your counters are just fine and there's a bug in OSHI. :|

I think localization may be necessary for that call (wildcard counters) but not for the other one and I overzealously fixed #1662.

**Reply by dbwiddis** [reply]
Can you please test the current SNAPSHOT build?   (Deployed from a local fix.). 

I think #1662 was an incorrect fix and this should correct it.

**Reply by unknown-spirit** [reply]
I already did and it says 
`"WARN oshi.util.platform.windows.PerfCounterWildcardQuery - Failed to locate performance object for Processor Information in the registry. Performance counters may be corrupt. Pdh call failed with error code 0xC0000BB8"`
and then falling back to WMI, as expected.

**Reply by dbwiddis** [reply]
That's not the latest snapshot.  I just built one from my local machine and it should have the localized name in that error message if it failed.

**Reply by dbwiddis** [reply]
Snapshot is based on changes now in PR #2089

**Reply by dbwiddis** [reply]
Which would have given an NPE.  So, NOW the latest snapshot.   Sorry :)

**Reply by unknown-spirit** [reply]
> Which would have given an NPE. So, NOW the latest snapshot. Sorry :)

Ok, this time it works. no errors and just the standard warning
`[main] WARN oshi.util.FileUtil - Failed to load default configuration`

**Reply by dbwiddis** [reply]
OK, so I think finally after 4 years all the localization bugs in the perf counters are fixed.

Interested in whether this solves your CPU load issues now.

**Reply by dbwiddis** [reply]
Also, I'm beyond thankful for your patience in helping me to run this to ground.  You have been EXCEEDINGLY helpful.

**Reply by dbwiddis** [reply]
Although I wonder which configuration file it's not reading :(

**Reply by unknown-spirit** [reply]
>Interested in whether this solves your CPU load issues now.

We will see. Having run it for ten minutes, I think it's equal or even slightly higher. On the other hand, the issue with the system process may be solved. This will take more time to investigate. Tomorrow is another day.

>Also, I'm beyond thankful for your patience in helping me to run this to ground. You have been EXCEEDINGLY helpful.

No problem. I just tested, your knowledge was the key. I know that localization of performance counters is a pain in the ass as I have used them too.

**Reply by unknown-spirit** [reply]
@dbwiddis
After much testing I can confirm two major observations:
1. The CPU usage of BusyTasks stays low for many hours.
2. The CPU usage of "NTOSKRNL.exe%Ordinal25" steadily increases and only restores on BusyTasks' exit, which strongly suggests a connection between the two processes.

>I suspect you're still seeing an impact from Garbage Collection, but not necessarily from OSHI. Where are all the byte[] allocations coming from?

I always thought that java manages memory on its own and you could safely rely on it, but that's obviously not the case.
It's clear that there is something wrong, be it in my program or even in MS Windows. Looking at my program, I can't extract anything reasonable from the huge number of profiler values, and regarding ntoskrnl, there's obviously no profiling option (access violation).
Do you have any tips on how to solve this issue? Is it wise to set unused Java-Pointers to null, even if a concrete assignment already suggests freeing of the old stuff?

**Reply by dbwiddis** [reply]
> Do you have any tips on how to solve this issue? Is it wise to set unused Java-Pointers to null, even if a concrete assignment already suggests freeing of the old stuff?

Unfortunately none specific.   I do think, generally speaking, that the additional NTOSKRNL load is associated with native memory management in some way.  However, I'm not really sure it's a problem that needs to be solved if the usage is down in the 5%-10% range.

> I always thought that java manages memory on its own and you could safely rely on it

Generally speaking, yes (via the Garbage Collector) but when you are dealing with JNA you are dealing with native memory which is a completely different thing outside of the GC world.  It's essentially low level `malloc()` and `free()` calls.  The latest changes in JNA and OSHI moved the scheduling of those earlier (so more memory is kept free at any given time) but the total number of calls is still somewhat the same.   

At this point I think the best thing you can do is experiment with various java command line switches to try out different configurations and see if they have any impact.  Try increasing heap size, changing GC thresholds, different GCs, etc.

You can try [tracking native memory](https://docs.oracle.com/javase/8/docs/technotes/guides/troubleshoot/tooldescr007.html) as well.

It's possible there's a device driver on your system that OSHI has to query that is outdated and is the cause of the issue.

**Reply by unknown-spirit** [reply]
Thanks for your input.
After a few days' search, I was able to detect the problem. With the help of "Windows Assessment and Deployment Center" I found suspicious calls of "StartSystemThread" and "MiWalkPageTablesRecursively". At that time I already had in mind some poor thread creation, which was finally confirmed by copying the workflow of [this site](https://superuser.com/questions/1360097/how-to-control-cpu-usage-of-ntoskrnl-exemiwalkpagetablesrecursively). My program was spamming short-living threads, which all were collected by Windows as zombies. They made up 10k after a few hours, slowing everything down. This issue will be solved in the next release, coming in about a week.
As far as I see now, the CPU usage of the program itself stays very low. Lower than I thought it could be. Thanks for updating oshi, release 6.2.0 did the trick!

**Reply by dbwiddis** [reply]
Excellent!   Thanks for the update and glad to hear it's all resolved!
